doc/man/man1/falsecolor.1
.\" RCSid "$Id: falsecolor.1,v 1.16 2023/08/15 23:39:58 greg Exp $"
.TH "FALSECOLOR" "1" "12/12/11" "RADIANCE" ""
.SH "NAME"
falsecolor \- make a false color RADIANCE picture
.SH "SYNOPSIS"
.B falsecolor
[
.B "\-i input"
][
.B "\-p picture"
][
.B "\-cb | \-cl | \-cp | \-c0"
][
.B \-e
][
.B "\-s scale"
][
.B "\-d digits"
][
.B "\-l label"
][
.B "\-n ndivs"
][
.B "\-lw lwidth"
][
.B "\-lh lheight"
][
.B "\-log decades"
][
.B "\-m mult"
][
.B "\-odim ncols nrows
][
.B "\-orct xmin ymin xmax ymax
][
.B "\-pal palette"
][
.B "\-r redv"
][
.B "\-g grnv"
][
.B "\-b bluv"
]

.B falsecolor \-palettes
.SH "DESCRIPTION"
.I Falsecolor
produces a false color picture for lighting analysis.
Input is a rendered Radiance picture.
.PP 
By default, luminance is displayed on a linear scale from 0 to 1000 cd/m2, where
dark areas are purple and brighter areas move through blue, green, red to yellow.
A different scale can be given with the
.I \-s
option.
If the argument given to
.I \-s
begins with an "a" for "auto," then the maximum is used for scaling the result.
The default multiplier is 179, which converts from radiance or irradiance
to luminance or illuminance, respectively.
A different multiplier can be given with
.I \-m
to get daylight factors or whatever.
For a logarithmic rather than a linear mapping, the
.I \-log
option can be used, where
.I decades
is the number of decades below the maximum scale desired.
.PP 
The 
.I \-d 
option controls the max number of decimal places printed for legend
entries. The default is 3 decimal places.
.PP 
A legend is produced for the new image with a label given by the
.I \-l
option.
The default label is "cd/m2", which is appropriate for standard Radiance
images.
If the
.I \-i
option of
.I rpict(1)
was used to produce the image, then the appropriate label would be "Lux".
.PP 
If contour lines are desired rather than just false color, the
.I \-cl
option can be used.
These lines can be placed over another Radiance picture using the
.I \-p
option.
If the input picture is given with
.I \-ip
instead of
.I \-i,
then it will be used both as the source of values and as the picture
to overlay with contours.
The
.I \-cb
option produces contour bands instead of lines, where the thickness of
the bands is related to the rate of change in the image.
The
.I \-cp
option creates a posterization effect where colours are banded without
the background image showing through.
The
.I \-n
option can be used to change the number of contours (and corresponding
legend entries) from the default value of 8.
The
.I \-lw
and
.I \-lh
options may be used to change the legend dimensions from the default width
and height of 100x200.
A value of zero for either dimension eliminates the legend in the output.
.PP
The
.I \-c0
option turns off false color output entirely, and is really only useful
with the
.I \-odim
option to overlay a matrix of values, or the
.I \-e
option to print the extrema values, as described below.
The
.I \-c0
option simultaneously removes the legend.
.PP
The
.I \-odim
option sets the number of columns and rows to overlay
a value matrix on the false color output.
The
.I \-orct
option may be used to specify a subrectangle of the image to overlay,
where the minimum x and y picture coordinates for the lower-left
corner are given before the maximum coordinates corresponding
to the upper-right corner.
The input picture must still cover the entire image area and match
the underlying background dimensions.
.PP
The
.I \-e
option causes extrema points to be printed on the brightest and
darkest pixels of the input picture, on top of any matrix values.
.PP 
The
.I "\-pal"
option provides different color palettes for
.I falsecolor.
The current choices are
.I spec
for the old spectral mapping,
.I tbo
for the "turbo" spectral mapping,
.I hot
for a thermal scale,
.I eco
for a blue-red-yellow scale, and
.I pm3d
for a variation of the default mapping,
.I def.
A Radiance HDR image of all available palettes can be created with the 
.I \-palettes
option.
The remaining options,
.I "\-r, \-g,"
and
.I \-b
are for changing the mapping of values to colors.
These are expressions of the variable 
.I v,
where
.I v
varies from 0 to 1.
These options are not recommended for the casual user.
.PP 
If no
.I \-i
or
.I \-ip
option is used, input is taken from the standard input.
The output image is always written to standard output, which should
be redirected, and is frequently converted to a non-HDR image format.
.SH "EXAMPLES"
To create a false color image directly from
.I rpict(1):
.IP "" .2i
rpict \-vf default.vp scene.oct | falsecolor > scene.hdr
.PP
To show the available color palettes:
.IP "" .2i
falsecolor -palettes | ximage
.PP 
To create a logarithmic contour plot of illuminance values on a
Radiance image:
.IP "" .2i
rpict -i -vf default.vp scene.oct > irrad.hdr
.br 
rpict -vf default.vp scene.oct > rad.hdr
.br 
falsecolor -i irrad.hdr -p rad.hdr -cl -log 2 -l Lux > lux.hdr
.PP
To overlay a matrix of 3 columns by 4 rows on the image above:
.IP "" .2i
falsecolor -odim 3 4 -i irrad.hdr -p rad.hdr -cl -log 2 -l Lux > lux.hdr
.PP
The same but without false color lines:
.IP "" .2i
falsecolor -odim 3 4 -i irrad.hdr -p rad.hdr -c0 > lux.hdr
.SH "AUTHOR"
Greg Ward
.br 
Axel Jacobs (Perl port and -pal options)
.br
Taoning Wang added the "turbo" palette.
.br
David Geisler-Moroder & Stephen Wasilewski (Legend additions and improvements)
.SH "ACKNOWLEDGEMENT"
Work on this program was initiated and sponsored by the LESO
group at EPFL in Switzerland.
The 'eco' palette was sponsored by Foster + Partners in London.
.SH "SEE ALSO"
getinfo(1), pcomb(1), pcompos(1), pextrem(1), pfilt(1), pflip(1), protate(1),
psign(1), pvalue(1), ra_tiff(1), ra_bmp(1), rpict(1), ximage(1)


src/common/falsecolor.c
#ifndef lint
static const char RCSid[] = "$Id: falsecolor.c,v 3.11 2021/01/07 01:28:36 greg Exp $";
#endif
/*
 * False color mapping functions.
 * See falsecolor.h for detailed function descriptions.
 *
 * Externals declared in falsecolor.h
 */

#include "copyright.h"

#include	<stdio.h>
#include	<math.h>
#include	<string.h>
#include	"tmprivat.h"
#include	"falsecolor.h"

/* Initialize new false color mapping */
FCstruct *
fcInit(uby8 fcscale[256][3])
{
	FCstruct	*fcs = (FCstruct *)malloc(sizeof(FCstruct));
	
	if (fcs == NULL)
		return(NULL);
	fcs->mbrmin = 10; fcs->mbrmax = -10;
	fcs->lumap = NULL;
	if ((fcs->scale = fcscale) == NULL)
		fcs->scale = fcDefaultScale;
	return(fcs);
}

/* Assign fixed linear false color map */
int
fcFixedLinear(FCstruct *fcs, double Lwmax)
{
	double	mult;
	int	i;

	if ((fcs == NULL) | (Lwmax <= MINLUM))
		return(TM_E_ILLEGAL);
	if (fcs->lumap != NULL)
		free((void *)fcs->lumap);
	fcs->mbrmin = tmCvLuminance(Lwmax/256.);
	fcs->mbrmax = tmCvLuminance(Lwmax);
	fcs->lumap = (uby8 *)malloc(sizeof(uby8)*(fcs->mbrmax - fcs->mbrmin + 1));
	if (fcs->lumap == NULL)
		return(TM_E_NOMEM);
	mult = 255.999/tmLuminance(fcs->mbrmax);
	for (i = fcs->mbrmin; i <= fcs->mbrmax; i++)
		fcs->lumap[i - fcs->mbrmin] = (int)(mult * tmLuminance(i));
	returnOK;
}

/* Assign fixed logarithmic false color map */
int
fcFixedLog(FCstruct *fcs, double Lwmin, double Lwmax)
{
	int	i;

	if ((fcs == NULL) | (Lwmin <= MINLUM) | (Lwmax <= Lwmin))
		return(TM_E_ILLEGAL);
	if (fcs->lumap != NULL)
		free((void *)fcs->lumap);
	fcs->mbrmin = tmCvLuminance(Lwmin);
	fcs->mbrmax = tmCvLuminance(Lwmax);
	if (fcs->mbrmin >= fcs->mbrmax) {
		fcs->lumap = NULL;
		return(TM_E_ILLEGAL);
	}
	fcs->lumap = (uby8 *)malloc(sizeof(uby8)*(fcs->mbrmax - fcs->mbrmin + 1));
	if (fcs->lumap == NULL)
		return(TM_E_NOMEM);
	for (i = fcs->mbrmax - fcs->mbrmin; i >= 0; i--)
		fcs->lumap[i] = 256L * i / (fcs->mbrmax - fcs->mbrmin + 1);
	returnOK;
}

/* Compute linear false color map */
int
fcLinearMapping(FCstruct *fcs, TMstruct *tms, double pctile)
{
	int		i, histlen;
	HIST_TYP	histot;
	long		cnt;
	int		brt0;

	if ((fcs == NULL) | (tms == NULL) || (tms->histo == NULL) |
			(0 > pctile) | (pctile >= 50))
		return(TM_E_ILLEGAL);
	i = HISTI(tms->hbrmin);
	brt0 = HISTV(i);
	histlen = HISTI(tms->hbrmax) + 1 - i;
	histot = 0;
	for (i = histlen; i--; )
		histot += tms->histo[i];
	cnt = histot * pctile / 100;
	for (i = histlen; i--; )
		if ((cnt -= tms->histo[i]) < 0)
			break;
	if (i <= 0)
		return(TM_E_TMFAIL);
	return(fcFixedLinear(fcs, tmLuminance(brt0 + i*HISTEP)));
}

/* Compute logarithmic false color map */
int
fcLogMapping(FCstruct *fcs, TMstruct *tms, double pctile)
{
	int		i, histlen;
	HIST_TYP	histot;
	long		cnt;
	int		brt0, wbrmin, wbrmax;

	if ((fcs == NULL) | (tms == NULL) || (tms->histo == NULL) |
			(.0 > pctile) | (pctile >= 50.))
		return(TM_E_ILLEGAL);
	i = HISTI(tms->hbrmin);
	brt0 = HISTV(i);
	histlen = HISTI(tms->hbrmax) + 1 - i;
	histot = 0;
	for (i = histlen; i--; )
		histot += tms->histo[i];
	cnt = histot * pctile * .01;
	for (i = 0; i < histlen; i++)
		if ((cnt -= tms->histo[i]) < 0)
			break;
	if (i >= histlen)
		return(TM_E_TMFAIL);
	wbrmin = brt0 + i*HISTEP;
	cnt = histot * pctile * .01;
	for (i = histlen; i--; )
		if ((cnt -= tms->histo[i]) < 0)
			break;
	wbrmax = brt0 + i*HISTEP;
	if (wbrmax <= wbrmin)
		return(TM_E_TMFAIL);
	return(fcFixedLog(fcs, tmLuminance(wbrmin), tmLuminance(wbrmax)));
}
 
/* Apply false color mapping to pixel values */
int
fcMapPixels(FCstruct *fcs, uby8 *ps, TMbright *ls, int len)
{
	int	li;

	if (fcs == NULL || (fcs->lumap == NULL) | (fcs->scale == NULL))
		return(TM_E_ILLEGAL);
	if ((ps == NULL) | (ls == NULL) | (len < 0))
		return(TM_E_ILLEGAL);
	while (len--) {
		if ((li = *ls++) < fcs->mbrmin)
			li = 0;
		else if (li > fcs->mbrmax)
			li = 255;
		else
			li = fcs->lumap[li - fcs->mbrmin];
		*ps++ = fcs->scale[li][RED];
		*ps++ = fcs->scale[li][GRN];
		*ps++ = fcs->scale[li][BLU];
	}
	returnOK;
}

/* Determine if false color mapping is logarithmic */
int
fcIsLogMap(FCstruct *fcs)
{
	int	miderr;

	if (fcs == NULL || fcs->lumap == NULL)
		return(-1);
	
	miderr = (fcs->mbrmax - fcs->mbrmin)/2;
	miderr = fcs->lumap[miderr] -
			256L * miderr / (fcs->mbrmax - fcs->mbrmin + 1);

	return((-1 <= miderr) & (miderr <= 1));
}

/* Duplicate a false color structure */
FCstruct *
fcDup(FCstruct *fcs)
{
	FCstruct	*fcnew;

	if (fcs == NULL)
		return(NULL);
	fcnew = fcInit(fcs->scale);
	if (fcnew == NULL)
		return(NULL);
	if (fcs->lumap != NULL) {
		fcnew->lumap = (uby8 *)malloc(sizeof(uby8)*(fcs->mbrmax -
							fcs->mbrmin + 1));
		if (fcnew->lumap == NULL)
			return(fcnew);
		fcnew->mbrmin = fcs->mbrmin; fcnew->mbrmax = fcs->mbrmax;
		memcpy((void *)fcnew->lumap, (void *)fcs->lumap,
				sizeof(uby8)*(fcs->mbrmax - fcs->mbrmin + 1));
	}
	return(fcnew);
}

/* Free data associated with the given false color mapping structure */
void
fcDone(FCstruct *fcs)
{
	if (fcs == NULL)
		return;
	if (fcs->lumap != NULL)
		free((void *)fcs->lumap);
	free((void *)fcs);
}

uby8	fcDefaultScale[256][3] = {	/* default false color scale */
	111,8,132,
	108,7,133,
	105,7,134,
	102,6,136,
	98,6,137,
	93,5,139,
	89,4,141,
	84,3,143,
	79,2,145,
	74,1,148,
	68,0,150,
	63,0,153,
	57,0,155,
	52,0,157,
	46,0,160,
	41,0,162,
	36,0,164,
	31,0,166,
	26,0,168,
	22,0,170,
	18,0,172,
	14,2,174,
	11,4,175,
	 8,7,176,
	 7,9,177,
	 6,11,177,
	 5,13,178,
	 4,16,178,
	 3,18,179,
	 2,21,180,
	 1,24,180,
	 1,28,181,
	 0,31,181,
	 0,35,182,
	 0,38,182,
	 0,42,183,
	 0,46,184,
	 0,50,184,
	 0,54,184,
	 0,58,185,
	 0,63,185,
	 0,67,186,
	 0,71,186,
	 0,76,186,
	 0,80,187,
	 0,84,187,
	 0,89,187,
	 0,93,187,
	 1,97,187,
	 1,102,187,
	 1,106,187,
	 2,110,187,
	 2,114,187,
	 3,118,186,
	 3,122,186,
	 4,126,186,
	 4,130,185,
	 4,133,185,
	 5,137,184,
	 5,140,183,
	 6,143,182,
	 6,146,181,
	 6,149,180,
	 7,151,179,
	 7,154,178,
	 7,156,177,
	 8,158,175,
	 8,161,172,
	 9,163,169,
	 9,165,165,
	 9,167,161,
	 9,169,157,
	10,170,153,
	10,172,148,
	10,173,143,
	11,174,138,
	11,174,133,
	11,175,127,
	12,175,122,
	12,176,117,
	13,176,111,
	14,176,106,
	14,176,101,
	15,175,95,
	16,175,90,
	17,175,86,
	18,174,81,
	20,174,77,
	21,173,73,
	22,172,69,
	24,172,66,
	26,171,63,
	28,170,60,
	30,169,58,
	32,168,57,
	34,167,56,
	37,166,55,
	40,165,54,
	42,164,54,
	45,163,54,
	48,162,55,
	52,160,55,
	55,158,56,
	58,157,57,
	62,155,57,
	66,153,59,
	69,152,60,
	73,150,61,
	77,148,63,
	81,146,64,
	84,144,66,
	88,142,67,
	92,139,69,
	96,137,70,
	99,135,72,
	103,133,73,
	107,131,75,
	110,128,76,
	113,126,77,
	117,124,78,
	120,121,79,
	123,119,80,
	126,117,80,
	128,114,81,
	131,112,81,
	133,110,81,
	135,108,80,
	136,106,80,
	137,105,80,
	138,104,79,
	139,102,79,
	140,101,79,
	141,100,78,
	142,98,78,
	143,96,77,
	144,95,76,
	144,93,76,
	145,92,75,
	146,90,74,
	146,89,73,
	147,87,73,
	148,85,72,
	148,84,71,
	149,82,70,
	149,80,69,
	150,79,68,
	150,77,67,
	151,75,66,
	151,73,65,
	151,72,64,
	152,70,63,
	152,68,62,
	153,66,61,
	153,65,60,
	153,63,59,
	154,61,58,
	154,60,57,
	154,58,56,
	154,56,55,
	155,55,54,
	155,53,53,
	155,51,51,
	156,50,50,
	156,48,49,
	156,46,48,
	157,45,47,
	157,43,46,
	157,42,45,
	158,40,44,
	158,39,43,
	158,37,42,
	159,36,41,
	159,34,40,
	159,33,39,
	160,32,38,
	160,31,37,
	161,29,37,
	161,28,36,
	162,27,35,
	162,26,34,
	163,25,33,
	163,24,33,
	164,23,32,
	165,22,31,
	165,21,31,
	168,18,29,
	170,16,28,
	172,13,26,
	175,11,25,
	177,9,24,
	180,7,23,
	183,5,22,
	185,3,21,
	188,2,21,
	191,1,20,
	194,0,19,
	197,0,19,
	199,0,18,
	202,0,17,
	205,0,17,
	207,0,16,
	210,2,16,
	213,3,15,
	215,6,14,
	217,8,13,
	219,11,13,
	220,13,12,
	222,17,11,
	224,20,11,
	226,24,10,
	227,28,9,
	229,32,8,
	231,37,7,
	232,42,6,
	234,47,5,
	236,52,5,
	237,57,4,
	239,63,3,
	240,68,2,
	242,74,2,
	243,79,1,
	245,85,0,
	246,91,0,
	247,96,0,
	248,102,0,
	250,108,0,
	251,113,0,
	252,118,0,
	253,123,0,
	254,128,0,
	254,133,0,
	255,138,0,
	255,143,1,
	255,148,2,
	255,154,3,
	255,159,4,
	255,165,6,
	255,170,7,
	255,176,9,
	255,181,11,
	255,187,13,
	255,192,15,
	255,198,17,
	255,203,20,
	255,208,22,
	255,213,24,
	255,218,26,
	255,223,28,
	255,227,30,
	255,232,32,
	255,236,34,
	254,240,35,
	254,243,37,
	254,246,38,
	254,249,39,
	254,252,40,
};
